Add workspace parser and database generation functionality

Javagedes opened the pull request.

makubacki commented:
I plan to look at this Tuesday/Wednesday of this week.

cfernald commented on edk2toollib/database/tables/source_table.py:
```diff
@@ -0,0 +1,72 @@
+        src_table = db.table("source", cache_size=None)
+
+        start = time.time()
+        files = list(ws.rglob("*.c")) + list(ws.rglob("*.h"))
```
Should this include cpp now that they are used for googletest? 

Also is assembly intentionally excluded?

Javagedes commented:
@cfernald 

Thank you for the review. I have updated the PR with your suggestions. Notably:

1. Users now set the Mode they wish to open in `Edk2DB(Edk2DB.MEM_RW, ... )`
2. I have added a Readme at docs/user/features/edk2_db.md that will be available in the public docs.

Let me know if you have any other questions or concerns!

Javagedes commented on docs/user/features/edk2_db.md:
```diff
@@ -0,0 +1,135 @@
+
+## Table Generators
+
+Table generators are just that, classes that subclass the [TableGenerator](/api/database/edk2_db/#edk2toollib.database.edk2_db.TableGenerator)
```
The link is fine. Technically it doesn't exist when you view it directly from the markdown, however this link is to be consumed when viewing this markdown file via tianocore.org/edk2-pytool-library. When we build the docs site, we auto-generate API references, which is what this link is referencing.

Javagedes commented on edk2toollib/database/tables/source_table.py:
```diff
@@ -0,0 +1,79 @@
+from edk2toollib.database import Edk2DB, TableGenerator
+
+
+SOURCE_FILES = ["*.c", "*.h", "*.cpp", "*.asm"]
```
Sure, I can add them all here for now. We may need to further parallelize (per file type) later when we do more then look for a license, but for now the speed of adding a few more file types don't appear to slow it down.
